src/GafferScene/CollectScenes.cpp

```diff
@@ -0,0 +1,430 @@
+
+using namespace std;
+using namespace Imath;
+using namespace IECore;
+using namespace Gaffer;
+using namespace GafferScene;
+
+//////////////////////////////////////////////////////////////////////////
+// Internal utilities
+//////////////////////////////////////////////////////////////////////////
+
+namespace
+{
+
+class SceneScope : public Context::EditableScope
+{
+
+	public :
+
+		SceneScope( const Context *context, const InternedString &rootNameVariable )
+			:	EditableScope( context ), m_rootNameVariable( rootNameVariable )
+		{
+		}
+
+		SceneScope( const Context *context, const InternedString &rootNameVariable, const ScenePlug::ScenePath &downstreamPath )
+			:	EditableScope( context ), m_rootNameVariable( rootNameVariable )
+		{
+			setScenePath( downstreamPath );
+		}
+
+		void setRootName( const InternedString &name )
+		{
+			set( m_rootNameVariable, name.string() );
+		}
+
+		void setScenePath( const ScenePlug::ScenePath &downstreamPath )
+		{
+			setRootName( downstreamPath[0] );
+			ScenePlug::ScenePath upstreamPath( downstreamPath.begin() + 1, downstreamPath.end() );
+			set( ScenePlug::scenePathContextName, upstreamPath );
+		}
+
+	private :
+
+		InternedString m_rootNameVariable;
+
+};
+
+} // namespace
+
+//////////////////////////////////////////////////////////////////////////
+// CollectScenes
+//////////////////////////////////////////////////////////////////////////
+
+IE_CORE_DEFINERUNTIMETYPED( CollectScenes );
+
+size_t CollectScenes::g_firstPlugIndex = 0;
+
+CollectScenes::CollectScenes( const std::string &name )
+	:	SceneProcessor( name, 1 )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+
+	addChild( new StringVectorDataPlug( "rootNames", Plug::In, new StringVectorData() ) );
+	addChild( new StringPlug( "rootNameVariable", Plug::In, "collect:rootName" ) );
+}
+
+CollectScenes::~CollectScenes()
+{
+}
+
+Gaffer::StringVectorDataPlug *CollectScenes::rootNamesPlug()
+{
+	return getChild<StringVectorDataPlug>( g_firstPlugIndex + 0 );
+}
+
+const Gaffer::StringVectorDataPlug *CollectScenes::rootNamesPlug() const
+{
+	return getChild<StringVectorDataPlug>( g_firstPlugIndex + 0 );
+}
+
+Gaffer::StringPlug *CollectScenes::rootNameVariablePlug()
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 1 );
+}
+
+const Gaffer::StringPlug *CollectScenes::rootNameVariablePlug() const
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 1 );
+}
+
+void CollectScenes::affects( const Gaffer::Plug *input, AffectedPlugsContainer &outputs ) const
+{
+	SceneProcessor::affects( input, outputs );
+
+	if( input == rootNamesPlug() )
+	{
+		outputs.push_back( outPlug()->childNamesPlug() );
+	}
+	else if( input == rootNameVariablePlug() )
+	{
+		for( PlugIterator it( outPlug() ); !it.done(); ++it )
+		{
+			outputs.push_back( it->get() );
+		}
+	}
+	else if( input->parent<ScenePlug>() == inPlug() )
+	{
+		outputs.push_back( outPlug()->getChild<Plug>( input->getName() ) );
+	}
+	else if( input == outPlug()->childNamesPlug() )
```

Comment:
Out of curiosity - what's the difference between having this block here instead of putting its contents into the block where the rootNamesPlug() is handled? Is that only because we don't want to trigger a recompute if the plug value changes from 'red green blue' to 'red green blue blue'?